### PR TITLE
Allow dynamic configuration of API environment and keys

### PR DESCRIPTION
**Pull request by nicolasblanco**
This PR is to allow dynamic configuration of API environment and keys. The default behaviour of reading from the global app config is kept.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+## master
+
+### Enhancements
+
+* Ability to optionally pass environment and API keys as options to all functions doing API calls. The default behaviour of reading from the global config is kept if those keys are not passed as arguments.
```
Comment (sorentwo) [inline CHANGELOG.md]: Please wrap this text at 80 characters (to match the rest of the CHANGELOG).

File: README.md [hunk 2]
```diff
@@ -47,6 +47,8 @@ config :braintree,
 Furthermore, the environment defaults to `:sandbox`, so you'll want to configure
 it with `:production` in `prod.exs`.
 
+You may optionally pass directly those configuration keys to all functions performing an API call. In that case, those keys will be used to perform the call.
```
Comment (sorentwo) [inline README.md]: Similarly, please wrap this at 80 to match the rest of the README.

File: lib/add_on.ex [hunk 3]
```diff
@@ -43,9 +43,9 @@ defmodule Braintree.AddOn do
 
       {:ok, addons} = Braintree.AddOns.all()
   """
-  @spec all() :: {:ok, [t]} | {:error, Error.t}
-  def all do
-    with {:ok, %{"add_ons" => add_ons}} <- HTTP.get("add_ons") do
+  @spec all(list | []) :: {:ok, [t]} | {:error, Error.t}
```
Comment (sorentwo) [inline lib/add_on.ex]: This spec (and the others) should use `Keyword.t` instead of `list | []`.

File: lib/http.ex [hunk 4]
```diff
@@ -84,18 +84,18 @@ defmodule Braintree.HTTP do
...
   ## Helper Functions
 
   @doc false
-  @spec build_url(binary) :: binary
-  def build_url(path) do
-    environment = Braintree.get_env(:environment, :sandbox)
-    merchant_id = Braintree.get_env(:merchant_id)
+  @spec build_url(binary, Keyword.t) :: binary
+  def build_url(path, opts) do
+    environment = opts[:environment] || Braintree.get_env(:environment, :sandbox)
```
Comment (sorentwo) [inline lib/http.ex]: Let's use `Keyword.get/3` for this instead of access (`[]`) and `||`.

File: lib/http.ex [hunk 5]
```diff
@@ -123,10 +123,10 @@ defmodule Braintree.HTTP do
   end
 
   @doc false
-  @spec build_headers() :: [tuple]
-  def build_headers do
-    public  = Braintree.get_env(:public_key)
-    private = Braintree.get_env(:private_key)
+  @spec build_headers(Keyword.t) :: [tuple]
+  def build_headers(opts) do
+    public  = opts[:public_key] || Braintree.get_env(:public_key)
```
Comment (sorentwo) [inline lib/http.ex]: Same here, let's use `Keyword.get/3`.

**Comment by nicolasblanco** [comment]
@sorentwo : Thanks! I've done your comments and rebased.

**Comment by sorentwo** [comment]
Thanks for this! There are a few integration test errors which I fixed up locally. I also attributed both of you within the CHANGELOG.

**Pull request closed without merging**